docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+
+### Compressible Resource Guarantees
+- For now, we are only supporting CPU.
+- Jobs that don’t have a request specified, will have a request of 0.
+- Containers are guaranteed to get the amount of CPU they request, they may or may not get additional CPU time (depending on the other jobs running).
+- Excess CPU resources will be distributed based on the amount of CPU requested. For example, suppose container A requests for 60% of the CPU, and container B requests for 30% of the CPU. Suppose that both containers are trying to use as much CPU as they can. Then the extra 10% of CPU will be distributed to A and B in a 2:1 ratio.
+- Containers will be throttled if they exceed their limit. If limit is unspecified, then the containers can use excess resources when available.
+
+### Incompressible Resource Guarantees
+- For now, we are only supporting memory.
+- Jobs that don’t have a request specified, will have a request of 0.
+- Containers are guaranteed to get the amount of memory they request, if they exceed their memory request, they could be killed (if some other job needs memory), but if they consume fewer resources than requested, they will not be killed.
+- Containers will be killed if they use more memory than their limit.
+
+### Tier Policy
+- *Best-Effort* pods are pods where every container’s request, for both memory and CPU, is 0 or unspecified. These pods are not requesting for any resource guarantees, and will be treated as lowest priority.
+- *Guaranteed* pods are pods where every container, for both memory and CPU, has the same request and limit, and request is non-zero for every container. These pods ask for a well-defined amount of resources and are therefore considered top-priority.
+- All other pods are *Burstable* - middle priority pods that have some form of minimal resource guarantee, but can use more resources when available.
+- A container than specifies a limit of 0 should be considered invalid.
+
+Note that the policy of classifying pods is for the most part orthogonal to the admission policy (described next), and can be easily tweaked.
+
+### Admission Policy
+- Pods will be admitted by Kubelet based on the sum of requests of its containers. The Kubelet will ensure that sum of requests of all containers (over all pods) is within the system’s resource (for both memory and CPU).
+- One big guiding principle is that Best-Effort pods should be “invisible” to all other pods. The main purpose of Best-Effort pods is to utilize resources not used by higher-tier pods. Consequently, Guaranteed and Burstable pods will pre-empt Best-Effort pods if necessary.

```

Comment:
@bgrant0607 Yeah I agree that pre-emption is a separate issue. I'll remove comments about preemption from the doc. As you say, Resource QoS will just be responsible for slowing down or killing tasks.